### PR TITLE
Add an observer that will trigger after service starts

### PR DESCRIPTION
**Pull request by alsamylkin**
This observer is useful if we want to call some function after service starts, e.g. yarpc client.

### CODE REVIEW

File: service/service.go [hunk 1]
```diff
@@ -146,3 +146,19 @@ func New(options ...Option) (Owner, error) {
...
 	return svc, nil
 }
+
+type niladicStart func()
+
+func (n niladicStart) OnInit(service Host) error      { return nil }
+func (n niladicStart) OnShutdown(reason Exit)         {}
+func (n niladicStart) OnCriticalError(err error) bool { return true }
+func (n niladicStart) OnStateChange(old State, new State) {
+	if old == Starting && new == Running {
```
Comment (sectioneight) [inline service/service.go]: Don't we transition through Starting and then running? E.g. won't `n` be called twice? Also, don't shadow `new`

Comment (alsamylkin) [inline service/service.go]: If n called twice, the test should panic..

File: service/service_test.go [hunk 2]
```diff
@@ -137,3 +139,22 @@ var validServiceConfig = map[string]interface{}{
...
+			loggingCore: loggingCore{
+				log: ulog.NopLogger,
+			},
+		},
+		observer: AfterStart(func() {
+			wg.Done()
+		}),
+	}
+
+	h.StartAsync()
```
Comment (madhuravi) [inline service/service_test.go]: Add tests to trigger all functions.

File: service/service.go [hunk 3]
```diff
@@ -146,3 +146,19 @@ func New(options ...Option) (Owner, error) {
 
 	return svc, nil
 }
+
+type niladicStart func()
```
Comment (madhuravi) [inline service/service.go]: Is niladic a well known word? Slight preference for a simpler word for better readability.

Comment (alsamylkin) [inline service/service.go]: Got it from https://golang.org/pkg/sync/#Once.Do
Niladic - zero arguments,
Monadic - one argument,
Dyadic...

File: service/service_test.go (hunk 2 above)
Comment (alsamylkin) [inline service/service_test.go]: This is just silly :)
I don't think we should test empty functions or functions always returning nils/true..

File: service/service.go (hunk 1 above)
Comment (alsamylkin) [inline service/service.go]: We do transition from [Starting](https://github.com/uber-go/fx/blob/master/service/host.go#L235) to [Running](https://github.com/uber-go/fx/blob/master/service/host.go#L297) in start

File: service/service.go [hunk 4]
```diff
@@ -146,3 +146,19 @@ func New(options ...Option) (Owner, error) {
...
+
+func (n niladicStart) OnInit(service Host) error      { return nil }
+func (n niladicStart) OnShutdown(reason Exit)         {}
+func (n niladicStart) OnCriticalError(err error) bool { return true }
+func (n niladicStart) OnStateChange(old State, curr State) {
+	if old == Starting && curr == Running {
```
Comment (sectioneight) [inline service/service.go]: Oooh, I read this as `||` for some reason which is why I was wondering about the double-trigger. 🤦‍♂️

File: service/service.go (hunk 3 above)
Comment (glibsm) [inline service/service.go]: we tend to call things like this `nop`, nope?

Comment (sectioneight) [inline service/service.go]: It's not a noop though. A noop is a stub, this is a function with zero arguments that **does** get called.

**Pull request merged**